// @embroider/core/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var package_1 = require("./package");
exports.Package = package_1.default;
var template_compiler_1 = require("./template-compiler");
exports.TemplateCompiler = template_compiler_1.default;
var options_1 = require("./options");
exports.optionsWithDefaults = options_1.optionsWithDefaults;
var to_broccoli_plugin_1 = require("./to-broccoli-plugin");
exports.toBroccoliPlugin = to_broccoli_plugin_1.default;
var package_cache_1 = require("./package-cache");
exports.PackageCache = package_cache_1.default;
var package_name_1 = require("./package-name");
exports.packageName = package_name_1.default;
var wait_for_trees_1 = require("./wait-for-trees");
exports.WaitForTrees = wait_for_trees_1.default;
var build_stage_1 = require("./build-stage");
exports.BuildStage = build_stage_1.default;
var get_or_create_1 = require("./get-or-create");
exports.getOrCreate = get_or_create_1.getOrCreate;
var js_handlebars_1 = require("./js-handlebars");
exports.jsHandlebarsCompile = js_handlebars_1.compile;
var app_1 = require("./app");
exports.AppBuilder = app_1.AppBuilder;
var messages_1 = require("./messages");
exports.todo = messages_1.todo;
exports.unsupported = messages_1.unsupported;
exports.warn = messages_1.warn;
exports.debug = messages_1.debug;
exports.expectWarning = messages_1.expectWarning;
exports.throwOnWarnings = messages_1.throwOnWarnings;
var paths_1 = require("./paths");
exports.explicitRelative = paths_1.explicitRelative;
var babel_filter_1 = require("./babel-filter");
exports.babelFilter = babel_filter_1.default;
