// @glimmer/syntax/dist/modules/es5/lib/traversal/walker.js
var Walker =
/*#__PURE__*/
function () {
  function Walker(order) {
    this.order = order;
    this.stack = [];
  }

  var _proto = Walker.prototype;

  _proto.visit = function visit(node, callback) {
    if (!node) {
      return;
    }

    this.stack.push(node);

    if (this.order === 'post') {
      this.children(node, callback);
      callback(node, this);
    } else {
      callback(node, this);
      this.children(node, callback);
    }

    this.stack.pop();
  };

  _proto.children = function children(node, callback) {
    var type;

    if (node.type === 'Block' || node.type === 'Template' && visitors.Program) {
      type = 'Program';
    } else {
      type = node.type;
    }

    var visitor = visitors[type];

    if (visitor) {
      visitor(this, node, callback);
    }
  };

  return Walker;
}();

export { Walker as default };
var visitors = {
  Program: function Program(walker, node, callback) {
    for (var i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },
  Template: function Template(walker, node, callback) {
    for (var i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },
  Block: function Block(walker, node, callback) {
    for (var i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },
  ElementNode: function ElementNode(walker, node, callback) {
    for (var i = 0; i < node.children.length; i++) {
      walker.visit(node.children[i], callback);
    }
  },
  BlockStatement: function BlockStatement(walker, node, callback) {
    walker.visit(node.program, callback);
    walker.visit(node.inverse || null, callback);
  }
};
