// @glimmer/util/dist/commonjs/es5/lib/platform-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keys = keys;
exports.unwrap = unwrap;
exports.expect = expect;
exports.unreachable = unreachable;
exports.exhausted = exhausted;
exports.symbol = exports.tuple = void 0;

function keys(obj) {
  return Object.keys(obj);
}

function unwrap(val) {
  if (val === null || val === undefined) throw new Error("Expected value to be present");
  return val;
}

function expect(val, message) {
  if (val === null || val === undefined) throw new Error(message);
  return val;
}

function unreachable(message) {
  if (message === void 0) {
    message = 'unreachable';
  }

  return new Error(message);
}

function exhausted(value) {
  throw new Error("Exhausted " + value);
}

var tuple = function tuple() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args;
};

exports.tuple = tuple;
var symbol = typeof Symbol !== 'undefined' ? Symbol : function (key) {
  return "__" + key + Math.floor(Math.random() * Date.now()) + "__";
};
exports.symbol = symbol;
