// @glimmer/util/dist/modules/es5/lib/collections.js
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

import { ensureGuid } from './guid';
export function dict() {
  return Object.create(null);
}
export function isDict(u) {
  return u !== null && u !== undefined;
}
export function isObject(u) {
  return typeof u === 'object' && u !== null;
}
export var DictSet =
/*#__PURE__*/
function () {
  function DictSet() {
    this.dict = dict();
  }

  var _proto = DictSet.prototype;

  _proto.add = function add(obj) {
    if (typeof obj === 'string') this.dict[obj] = obj;else this.dict[ensureGuid(obj)] = obj;
    return this;
  };

  _proto["delete"] = function _delete(obj) {
    if (typeof obj === 'string') delete this.dict[obj];else if (obj._guid) delete this.dict[obj._guid];
  };

  return DictSet;
}();
export var StackImpl =
/*#__PURE__*/
function () {
  function StackImpl() {
    this.stack = [];
    this.current = null;
  }

  var _proto2 = StackImpl.prototype;

  _proto2.push = function push(item) {
    this.current = item;
    this.stack.push(item);
  };

  _proto2.pop = function pop() {
    var item = this.stack.pop();
    var len = this.stack.length;
    this.current = len === 0 ? null : this.stack[len - 1];
    return item === undefined ? null : item;
  };

  _proto2.nth = function nth(from) {
    var len = this.stack.length;
    return len < from ? null : this.stack[len - from];
  };

  _proto2.isEmpty = function isEmpty() {
    return this.stack.length === 0;
  };

  _proto2.toArray = function toArray() {
    return this.stack;
  };

  _createClass(StackImpl, [{
    key: "size",
    get: function get() {
      return this.stack.length;
    }
  }]);

  return StackImpl;
}();
