// @glimmer/util/dist/commonjs/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  assertNever: true,
  EMPTY_ARRAY: true,
  assert: true,
  deprecate: true,
  dict: true,
  DictSet: true,
  isDict: true,
  isObject: true,
  Stack: true,
  ensureGuid: true,
  initializeGuid: true,
  isSerializationFirstNode: true,
  SERIALIZATION_FIRST_NODE_STRING: true,
  EMPTY_SLICE: true,
  LinkedList: true,
  ListNode: true,
  ListSlice: true,
  assign: true,
  fillNulls: true,
  values: true,
  debugToString: true
};
exports.assertNever = assertNever;
Object.defineProperty(exports, "EMPTY_ARRAY", {
  enumerable: true,
  get: function () {
    return _arrayUtils.EMPTY_ARRAY;
  }
});
Object.defineProperty(exports, "assert", {
  enumerable: true,
  get: function () {
    return _assert.default;
  }
});
Object.defineProperty(exports, "deprecate", {
  enumerable: true,
  get: function () {
    return _assert.deprecate;
  }
});
Object.defineProperty(exports, "dict", {
  enumerable: true,
  get: function () {
    return _collections.dict;
  }
});
Object.defineProperty(exports, "DictSet", {
  enumerable: true,
  get: function () {
    return _collections.DictSet;
  }
});
Object.defineProperty(exports, "isDict", {
  enumerable: true,
  get: function () {
    return _collections.isDict;
  }
});
Object.defineProperty(exports, "isObject", {
  enumerable: true,
  get: function () {
    return _collections.isObject;
  }
});
Object.defineProperty(exports, "Stack", {
  enumerable: true,
  get: function () {
    return _collections.StackImpl;
  }
});
Object.defineProperty(exports, "ensureGuid", {
  enumerable: true,
  get: function () {
    return _guid.ensureGuid;
  }
});
Object.defineProperty(exports, "initializeGuid", {
  enumerable: true,
  get: function () {
    return _guid.initializeGuid;
  }
});
Object.defineProperty(exports, "isSerializationFirstNode", {
  enumerable: true,
  get: function () {
    return _isSerializationFirstNode.isSerializationFirstNode;
  }
});
Object.defineProperty(exports, "SERIALIZATION_FIRST_NODE_STRING", {
  enumerable: true,
  get: function () {
    return _isSerializationFirstNode.SERIALIZATION_FIRST_NODE_STRING;
  }
});
Object.defineProperty(exports, "EMPTY_SLICE", {
  enumerable: true,
  get: function () {
    return _listUtils.EMPTY_SLICE;
  }
});
Object.defineProperty(exports, "LinkedList", {
  enumerable: true,
  get: function () {
    return _listUtils.LinkedList;
  }
});
Object.defineProperty(exports, "ListNode", {
  enumerable: true,
  get: function () {
    return _listUtils.ListNode;
  }
});
Object.defineProperty(exports, "ListSlice", {
  enumerable: true,
  get: function () {
    return _listUtils.ListSlice;
  }
});
Object.defineProperty(exports, "assign", {
  enumerable: true,
  get: function () {
    return _objectUtils.assign;
  }
});
Object.defineProperty(exports, "fillNulls", {
  enumerable: true,
  get: function () {
    return _objectUtils.fillNulls;
  }
});
Object.defineProperty(exports, "values", {
  enumerable: true,
  get: function () {
    return _objectUtils.values;
  }
});
Object.defineProperty(exports, "debugToString", {
  enumerable: true,
  get: function () {
    return _debugToString.default;
  }
});

var _arrayUtils = require("./lib/array-utils");

var _assert = _interopRequireWildcard(require("./lib/assert"));

var _collections = require("./lib/collections");

var _destroy = require("./lib/destroy");

Object.keys(_destroy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _destroy[key];
    }
  });
});

var _dom = require("./lib/dom");

Object.keys(_dom).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dom[key];
    }
  });
});

var _guid = require("./lib/guid");

var _isSerializationFirstNode = require("./lib/is-serialization-first-node");

var _lifetimes = require("./lib/lifetimes");

Object.keys(_lifetimes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _lifetimes[key];
    }
  });
});

var _listUtils = require("./lib/list-utils");

var _objectUtils = require("./lib/object-utils");

var _platformUtils = require("./lib/platform-utils");

Object.keys(_platformUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _platformUtils[key];
    }
  });
});

var _string = require("./lib/string");

Object.keys(_string).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _string[key];
    }
  });
});

var _immediate = require("./lib/immediate");

Object.keys(_immediate).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _immediate[key];
    }
  });
});

var _template = require("./lib/template");

Object.keys(_template).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _template[key];
    }
  });
});

var _debugToString = _interopRequireDefault(require("./lib/debug-to-string"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function assertNever(value, desc) {
  if (desc === void 0) {
    desc = 'unexpected unreachable branch';
  }

  console.log('unreachable', value);
  console.trace(desc + " :: " + JSON.stringify(value) + " (" + value + ")");
}
